text/0000-katello-as-a-plugin.md

```diff
@@ -0,0 +1,107 @@
+
+## Smart Proxy vs. Capsule
+
+A Capsule is essentially a Smart Proxy that deploys Pulp, and a reverse proxy to achieve content federation and an isolated communication point fo clients. Foreman deploys the Smart Proxy on port 8443 by default, while Katello deploys it on 9090 by default. Katello deployed Smart Proxies on 9090 because Candlepin was already running on 8443 on the server.
+
+#### Design
+
+Given the above discussion of infrastructure port, introduce a designated Smart Proxy port (e.g. 9071) that is predictable and not commonly used as 8443 is today. Further, merge the puppet-capsule and puppet-foreman_proxy modules to unify Smart Proxies and Capsules and give the content aspects of a Smart Proxy a new name (e.g. content node or content proxy).
+
+#### Drawbacks
+
+The port change would require infrastructure propogation for existing users. The current Capsule concept does not use a Smart Proxy in the traditional sense given it uses a reverse proxy for clients to talk back to the server in an isolated way and deploys Pulp for content federation. Given Pulp has it's own stable API, there is no need to go through the Smart Proxy. However, the Smart Proxy is used to provide statusof Pulp, storage information and indicators of enabled features.
+
+## Puppet Modules
+
+Currently Katello maintains a set of puppet modules that set it a part from other plugins. In order to install and configure Katello, it requires `puppet-katello` for configuring Katello as a plugin and backend services and `puppet-capsule` to configure content features on a Smart Proxy. Further, Katello maintains a set of puppet modules for each of it's backend services including the aforementioned certificate setup.
+
+#### Design
+
+This has been previously discussed and outlined by an existing [foreman-dev post](https://groups.google.com/forum/#!topic/foreman-dev/hebniL0yyqM). The larger proposals would be:
+
+ * Add Katello plugin manifest to puppet-foreman
+ * Add Katello plugin manifests for each backend service to puppet-foreman
+ * Add puppet-capsule features to puppet-foreman_proxy
+
+#### Drawbacks
+
+Adds a lot more manifests to puppet-foreman and puppet-foreman_proxy and keeps things less separated and focused.
+
+## Installer Repositories
+
+The `katello-installer` uses the foreman-installer and scenarios in order to install providing it's own scenario for Katello and for the Capsule. These including the system checks, and hooks are all contained within the `katello-installer` repository.
+
+### Design
+
+Unify the installer by storing multiple scenarios and dependencies in the single `foreman-installer` repository. This would put everything in one place and encourage more design work to support a variety of scenarios for different deployment types.
+
+### Drawbacks
+
+Merging the two together could muddy the waters with respect to different sets of hooks, requirements and modules required to install a given scenario. 
```

Comment:
I could see the upgrade that katello does in hooks changed in one of two ways, 1) move all execute calls to puppet module where the rest is 2) extract to separate smart upgrade script that would be triggered from the hook

One advantage of 1) is that it could be used for upgrading katello using puppetmaster infrastructure, OTOH I think it does not help with recoveries from failed updates etc.

2) brings a way to have a better control of update flow, e.g. retry if something failed. It could be run independently but also from the installer run.